content/blog/2024/2024-06-06-junit5-functional-interfaces.md

```diff
@@ -0,0 +1,620 @@
+// other variants of assertTimeoutPreemptively accepting executables
+```
+
+Let's learn about these methods one by one.
+
+### Using Executables in `assertAll()`
+
+**The `Assertions.assertsAll()` method asserts that all supplied executables do not throw exceptions.**
+
+Let's now see how to use `assertAll()` and executables:
+
+```java
+public class ExecutableTest {
+  private List<Long> numbers = Arrays.asList(100L, 200L, 50L, 300L);
+  private Executable sorter =
+      () -> {
+        TimeUnit.SECONDS.sleep(2);
+        numbers.sort(Long::compareTo);
+      };
+  private Executable checkSorting =
+      () -> assertEquals(List.of(50L, 100L, 200L, 300L), numbers);
+  private Executable noChanges = 
+      () -> assertEquals(List.of(100L, 200L, 50L, 300L), numbers);
+  // tests
+}
+```
+
+In the `ExecutableTest` class, we define several tests to demonstrate the usage of the `Executable` functional interface with JUnit's timeout assertions.
+
+We start by initializing a list of `Long` numbers (numbers) and defining two `Executable` lambdas: `sorter` and `checkSorting`. The `sorter` lambda simulates a time-consuming operation by sleeping for 2 seconds and then sorting the list. The `checkSorting` lambda verifies that the list is in correct sort order. Additionally, we define another `Executable` lambda, `noChanges`, which checks that the list remains in its initial unsorted state.
+
+Consider following example that shows how to use `assertAll` with executable:
+
+```java
+@ParameterizedTest
+@CsvSource({"1,1,2,Hello,H,bye,2,byebye",
+            "4,5,9,Good,Go,Go,-10,",
+            "10,21,31,Team,Tea,Stop,-2,"})
+void testAssertAllWithExecutable(int num1, int num2, int sum,
+                                  String input, String prefix,
+                                  String arg, int count, String result) {
+  assertAll(
+      () -> assertEquals(sum, num1 + num2),
+      () -> assertTrue(input.startsWith(prefix)),
+      () -> {
+        if (count < 0) {
+          assertThrows(
+              IllegalArgumentException.class,
+              () -> {
+                new ArrayList<>(count);
+              });
+        } else {
+          assertEquals(result, arg.repeat(count));
+        }
+      });
+}
+```
+
+In `testAssertAllExecutable()` method, `Assertions.assertAll()` groups 3 assertions. The first checks if sum of `num1` and `num2` equals `sum`. The second verifies if string input starts with `prefix`. Finally, the third checks if `count` is less than 0, asserts that creating `ArrayList` with count will throw `IllegalArgumentException`, otherwise asserts that string `arg` repeated count times equals result.
+
+{{% info title="Lambda Expression Conversion Behind the Scenes" %}}
+
+Modern Java versions (starting from Java 8) use the `invokedynamic` instruction to handle lambda expressions more efficiently than creating anonymous classes.
+
+Here's an explanation of what happens behind the scenes:
```

Comment:
I think we do not need this as part of this article. There's an explanation of how lambda's work under the hood in your previous article. Here, I think, it's a bit off-topic. What do you think?

Reply:
Earlier we had not explained `invokedynamic`

Reply:
If I am not mistaken, you also covered `invokedynamic` in the article about functional interfaces. However, I still think it's a bit off-topic here as we focus on JUnit and not on JVM internals.

Reply:
Removed it.